src/stream/xml_writer.rs

```diff
@@ -157,81 +144,125 @@ impl<W: Write> XmlWriter<W> {
             Ok(())
         })
     }
+
+    fn handle_pending_collection(&mut self) -> Result<(), Error> {
+        if let Some(PendingCollection::Array) = self.pending_collection {
+            let res = self.write_value_event(EventKind::StartArray, |this| {
+                this.start_element("array")?;
+                this.stack.push(Element::Array);
+                Ok(())
+            });
+            self.pending_collection = None;
+            res
+        } else if let Some(PendingCollection::Dictionary) = self.pending_collection {
+            let res = self.write_value_event(EventKind::StartDictionary, |this| {
+                this.start_element("dict")?;
+                this.stack.push(Element::Dictionary);
+                Ok(())
+            });
+            self.pending_collection = None;
+            res
+        } else {
+            Ok(())
+        }
+    }
 }
 
 impl<W: Write> Writer for XmlWriter<W> {
     fn write_start_array(&mut self, _len: Option<u64>) -> Result<(), Error> {
-        self.write_value_event(EventKind::StartArray, |this| {
-            this.start_element("array")?;
-            this.stack.push(Element::Array);
-            Ok(())
-        })
+        self.handle_pending_collection()?;
+        self.pending_collection = Some(PendingCollection::Array);
+        Ok(())
     }
 
     fn write_start_dictionary(&mut self, _len: Option<u64>) -> Result<(), Error> {
-        self.write_value_event(EventKind::StartDictionary, |this| {
-            this.start_element("dict")?;
-            this.stack.push(Element::Dictionary);
-            Ok(())
-        })
+        self.handle_pending_collection()?;
+        self.pending_collection = Some(PendingCollection::Dictionary);
+        Ok(())
     }
 
     fn write_end_collection(&mut self) -> Result<(), Error> {
         self.write_event(|this| {
-            match (this.stack.pop(), this.expecting_key) {
-                (Some(Element::Dictionary), true) => {
+            match (
```

Comment:
This match is getting a bit hard to follow. I think you can just check beforehadn if there is a pending collection, and if so just write the empty tag and return before the existing match statement.

Reply:
split it up into two matches now